[2단계 - 자주 가는 음식점] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
## `Step2` 점심 뭐 먹지 미션
안녕하세요, 마르코! 니야입니다.
2단계 점심 뭐 먹지 미션 제출하겠습니다 :)

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: [니야의 ‘점심 뭐 먹지’ 웹사이트](https://sooyeoniya.github.io/javascript-lunch/)

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

✅ **RestaurantForm에서 value 값 객체로 매핑해 반환하는 메서드 리팩토링**

기존에는 `this.formElements`라는 요소를 가진 객체 자체를 `Object.entries()`를 사용하여 2차원 배열로 만들어 `reduce`를 통해 순회하며 해당 요소 내부에서 `querySelector`를 통해 `value` 값을 추출해주었습니다.

```jsx
// 타입 변환 전 코드
#getFormData() {
  return Object.entries(this.formElements).reduce((acc, [key, el]) => {
    acc[key] = el.querySelector("input, select, textarea").value;
    return acc;
  }, {});
}
```

해당 `value` 값들은 다음 `Restaurant` 인터페이스와 동일한 타입이기 때문에 이를 활용해주어야겠다는 생각이 들었습니다.

```tsx
export interface Restaurant {
  id: Uuid;
  category: Exclude<Category, "전체">;
  name: string;
  distance: Distance;
  description: string;
  link: UrlString;
  isFavorite: boolean;
}
```

근데 타입스크립트로 변환하는 과정에서 문제가 발생했습니다. 

여기서 `Object.entries(this.formElements)`의 `key` 타입이 그냥 `string`이 되어버렸습니다.

이 부분에 있어서 `this.formElements`의 정확한 `key` 타입은 `category | name | distance | description | link` 이었기 때문에 `FormElements`의 `key` 타입을 가져오기로 했습니다.

따라서 아래와 같은 타입을 정의하여 `Object.entries()` 결과 타입을 확실히 지정할 수 있도록 했습니다. 이렇게 하게 되면 객체 T의 각 키 K에 대해 튜플 타입으로 해서 `Object.entries()` 결과 타입과 동일하게 해주었습니다. (예를 들어 이런식으로: `[ ['category', HTMLDivElement], ['name', HTMLDivElement], … ]`)

```jsx
type Entries<T> = {
  [K in keyof T]: [K, T[K]];
}[keyof T][];
```

그리고 아래처럼 제네릭 타입 인자에 FormElements 를 넣어주고 아래와 같이 타입을 지정해주고, reduce의 초깃값을 `Omit<Restaurant, "id" | "isFavorite">` 으로 지정하여 그 타입대로 객체 프로퍼티가 설정되도록 했습니다.

```jsx
#getFormData() {
  return (Object.entries(this.formElements) as Entries<FormElements>).reduce(
    (acc, [key, el]) => {
      const query = el.querySelector("input, select, textarea");
      if (query !== null) {
        acc[key] = query.value;
      }
      return acc;
    },
    {} as Omit<Restaurant, "id" | "isFavorite">
  );
}
```

하지만 여기서 또 문제가 발생했는데요.

`reduce`를 사용하면서 `value` 값이 계속 any 타입으로 잡혀서 타입 오류가 난다는 것이었습니다.

`'Element' 형식에 'value' 속성이 없습니다.ts(2339)`

`acc`(`Omit<Restaurant, "id" | "isFavorite">`)와 `key`(`"category" | "name" | "distance" | "description" | "link”`)의 타입은 지정이 잘 되었는데, `Restaurant`의 각 `key`에 대한 `value`의 타입이 제대로 잡히지 않은 것 같았습니다.

따라서, 아예 구조를 바꿔야겠다고 생각했습니다.

사실 생각해보면 `this.formElements`는 `FormElements` 인터페이스에서 이미 정해진 구조이기 때문에 `Object.entries`와 `reduce`를 사용해 굳이 동적 접근할 필요가 없다는 생각이 들었습니다. 오히려 객체를 만드는 데 있어서, 인터페이스의 `key` 값 하나하나 접근하여 `value`를 가져오는 것이 더 효율적이라고 생각했습니다.

따라서 최종적으로 아래와 같이 수정하게 되었습니다!

이렇게 객체의 각 프로퍼티 반환 값을 `getInputValue`라는 동일한 메서드를 사용해 `key` 값만 넣어주어 재사용하도록 해주었습니다.

```jsx
#getFormData(): Omit<Restaurant, "id" | "isFavorite"> {
  return {
    category: this.#getInputValue("category"),
    name: this.#getInputValue("name"),
    distance: this.#getInputValue("distance"),
    description: this.#getInputValue("description"),
    link: this.#getInputValue("link"),
  };
}
```

`getInputValue`메서드에서는 `key` 타입을 `keyof Omit<Restaurant, "id" | "isFavorite">`로 지정해두고, 반환 타입을 `Restaurant[K]`로 해두었습니다.

그리고 해당 키 값에 대한 요소를 `this.formElements`에서 하나씩 가져와, 키와 선택자를 매핑한 객체인 `selectorMap`을 통해 해당 내부 요소에 접근하여 `value`를 추출해주었습니다. 이때 `el`의 타입 확인과 `value`의 반환 타입을 확실히 지정해주어야 타입 오류가 나지 않았습니다. `value`가 `undefined`인 경우에는 빈 문자열이 반환되도록 설정해주었습니다. (이 부분에 있어서는 `description`, `link`가 사용자 선택적 입력 값이긴 하지만, 아예 `undefined` 자체를 처리하는 것보다 빈 문자열로 지정하여 값을 처리하는 것이 더 낫다고 판단했기 때문입니다.)

```jsx
#getInputValue<K extends keyof Omit<Restaurant, "id" | "isFavorite">>(
  key: K
): Restaurant[K] {
  const element = this.formElements[key];
  if (!element) {
    return "" as Restaurant[K];
  }

  const selectorMap = {
    category: "select",
    name: "input",
    distance: "select",
    description: "textarea",
    link: "input",
  } as Record<K, string>;

  const el = element.querySelector(selectorMap[key]);

  if (
    el instanceof HTMLInputElement ||
    el instanceof HTMLSelectElement ||
    el instanceof HTMLTextAreaElement
  ) {
    return (el.value ?? "") as Restaurant[K];
  }

  return "" as Restaurant[K];
}
```

✅ **옵저버 패턴을 활용한 리팩토링**

기존 1단계 코드에서 음식점 목록에 대한 상태 관리에 대해 옵저버 패턴을 적용했었는데요. 그때 제가 잘못 구현한 내용이 있었습니다. 리스너가 리스트를 구독자들에게 보내고 있었는데 그걸 사용하지 않았던 것입니다.

```jsx
// RestaurantStore.ts
subscribe(listener: Listeners) {
  this.#listeners.add(listener);
  return () => this.#listeners.delete(listener);
}
#notifyListeners() {
  this.#listeners.forEach((listener) => listener(this.#restaurants));
}

// App.ts
this.restaurantStore.subscribe(() => this.#updateRestaurantList());
```

여기서 구독해주는 것을 단순히 `updateRestaurantList`라는 메서드 호출만 해주고, 내부적으로는 전부다 `RestaurantStore`에 있는 `getRestaurants`를 그대로 가져와서 반환해 사용해주었습니다.

이렇게 되면 옵저버 패턴을 사용하는 의미가 없어서 좀 더 잘 활용해보는 방식으로 수정해봤습니다.

우선 `RestaurantService`는 삭제해주었습니다. 그 이유는 `RestaurantStore` 에 있는 메서드를 한 번 더 참조해서 가져와주는, 단순히 하나의 경로만 더 제공해주는 느낌이 있었습니다. 그리고 `RestaurantStore`에 있는 상태들이 해당 비즈니스로직과 깊게 결합하는 상태들이기 때문에 Service에서 따로 처리해주면 오히려 응집도가 떨어질 것 같다는 느낌이 들었습니다. 그래서 `RestaurantStore` 하나로 `Restaurant`에 대한 전체 도메인을 관리해주도록 했습니다.

그리고 해당 도메인에 총 4개의 상태를 관리해주도록 했습니다.

```tsx
restaurants: Restaurant[]; // 음식점 목록
filteredRestaurants: Restaurant[]; // 필터링된 음식점 목록
currentFilter: FilterOptions; // 현재 필터 옵션들
selectedRestaurant: Restaurant; // 현재 선택된 음식점
```

`restaurants`는 해당 도메인 내부에서 계속 사용되는 기준이 되는 음식점 목록 리스트로 로컬스토리지에 저장하고 받아오며, 여러 메서드에 따라 업데이트 되도록 하는 상태입니다.

그리고 `filteredRestaurants` 같은 경우에는 여러 필터링 값을 적용하여 필터링된 음식점 목록을 저장하는 값입니다. 이 값이 외부에서 업데이트될 때 사용되는 상태입니다. 왜냐하면 모든 화면에 출력되는 음식점 목록은 ‘모든 음식점’이냐, ‘자주 가는 음식점’이냐에 따라 다르고, 카테고리에 따라, 이름순/거리순과 같은 정렬 방식에 따라 다르기 때문입니다.

`currentFilter`는 현재 설정된 필터링 옵션들입니다. 기존에는 필터링되는 이벤트를 가진 각 컴포넌트의 내부에 있는 메서드를 App.js로 불러와서 업데이트 해주도록 했습니다. 이렇게 하면 매번 값을 바로 불러올 수 있다는 장점이 있긴 했는데, 상태 흐름이 너무 명확하지 않아서 알아보기 어려웠습니다. 그래서 아예 필터링 값들도 `Store`에서 관리해주도록 하여 필터링 옵션들이 변경될 때마다 음식점 목록이 해당 필터링 값으로 업데이트 되도록 해주었습니다.

`selectedRestaurant`의 경우에는 상세 정보 바텀 시트를 위해 저장해 둔 값입니다. `Store`에 있는 `toggleFavorite` 메서드를 통해 상태가 업데이트가 되었을 때, 기존에는 `getRestaurantInfo` 메서드를 통해 해당 음식점 `id`에 대한 정보를 직접 받아와서 업데이트 해주었는데요.

```jsx
onToggleFavorite: (restaurantId: Restaurant["id"]) => {
  this.restaurantService.toggleFavorite(restaurantId);
  const updatedInfo: Restaurant | undefined =
    this.restaurantService.getRestaurantInfo(restaurantId);
  if (updatedInfo) this.$restaurantDetail.openDetail(updatedInfo);
},
```

아예 상세 정보 바텀 시트 부분을 구독해두어 `toggleFavorite` 메서드로 정보가 업데이트 되는 경우 자동으로 현재 상세 정보 바텀 시트에 띄워져 있는 음식점 정보 상태가 업데이트 되도록 해주었습니다.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

이번에는 지난 미션보다 기능이 좀 더 많아져서 어떻게 하면 좀 더 가독성 좋고 효율적인 설계가 나올지 고민했던 것 같습니다.
지난 피드백처럼 전반적인 설계에 대해서 피드백 해주시면 감사할 것 같아요!
그리고 TS 관련해서도 수정할만한 부분이 있다면 피드백 부탁드립니다. 
UX 적으로 불편 사항이 있거나, 버그가 있다면 말씀부탁드립니다. 감사합니다 😃

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

sooyeoniya committed:
docs: 2단계 요구사항 README 문서 정리

sooyeoniya committed:
feat(RestaurantFilter): 음식점 목록 필터링 UI 로직 추가

sooyeoniya committed:
chore(RestaurantFilter): name 속성 명칭 변경 및 import 대소문자 수정

sooyeoniya committed:
feat(RestaurantNavBar): 음식점 리스트 탭 UI 로직 구현

sooyeoniya committed:
feat(RestaurantListItem): 자주 가는 음식점 버튼 UI 로직 구현

sooyeoniya committed:
feat: 로컬 스토리지에 음식점 리스트 저장 및 자주 가는 음식점 기능 구현

- 음식점 리스트 정보에 isFavorite 여부와 id 속성 추가하여 로컬 스토리지에 저장
- id 는 UUID를 생성해주는 유틸함수 구현하여 사용
- 현재 filterType에 따라 다르게 리스트 렌더링

sooyeoniya committed:
chore: 폼 내부 마진 줄이기

- 너무 입력 아이템 간의 공백이 커서 줄임

sooyeoniya committed:
refactor: 네비게이션바 키 상수 적용

- 미적용된 부분 마저 적용함

sooyeoniya committed:
docs: README.md 기능 구현한 내용 체크

sooyeoniya committed:
refactor(constants): Label Key 생성 및 Sort Option Text 객체 재정의

- Label Key 를 사용해서 Sort Option Text의 키 값 지정

sooyeoniya committed:
feat(RestaurantFilter): 자주가는 음식점 탭에서 음식점 필터 화면에서 제거

- 자주가는 음식점 탭에서는 음식점 필터하는 드롭다운 메뉴(카테고리, 정렬)가 보이면 안됨
- restaurantList의 update 메서드 이름 변경 (무엇을 업데이트하는지 메서드명에 정확히 명시)
- restaurantFilter에서 '전체' 요소 합쳐서 배열 따로 분리

sooyeoniya committed:
feat: 카테고리에 따른 정렬 로직 구현

- RestaurantFilter 내부 렌더링 로직 분리
- CATEGORY 상수에 '전체' 추가 및 적용
- 음식점 목록 업데이트 로직에 filterType 인자로 추가(filterType 에 카테고리 및 이름순/거리순 정렬 프로퍼티 추가)
- 카테고리에 따른 필터링 로직 구현

sooyeoniya committed:
feat: 메뉴 탭, 카테고리, 정렬 필터링 처리 구현

- 메뉴 탭, 카테고리, 이름순/거리순에 따른 음식점 리스트 정렬 및 필터링 처리 구현

sooyeoniya committed:
refactor: BottomSheetBase 내부 title을 RestaurantForm 으로 직접 주입하도록 수정

- BottomSheetBase 컴포넌트 재사용성 향상을 위함
- 상세정보 탭에서 BottomSheetBase 컴포넌트 재사용 필요

sooyeoniya committed:
feat: 음식점 상세 정보 구현

모달 열고 닫기, 즐겨찾기, 삭제 기능 추가
기존 바텀시트와 구분하기 위해 App에서 메서드명 변경
Button 스타일 변환에서 액션 추가 (close, delete)

sooyeoniya committed:
refactor(RestaurantDetail): 모달창이 다시 열릴 때 상세정보만 변경되도록 수정

- 이전 코드에서는 form부터 전부다 재생성되어 리렌더링이 일어났음
- 따라서 모달창이 열릴 때 변경되는 값들에 대한 요소들만 리렌더링 일어나도록 수정함

sooyeoniya committed:
feat(RestaurantDetail): 상세정보 탭에서 즐찾 버튼 클릭 시 바로 렌더링

- 추가로 RestaurantList 에서 restaurantService를 바로 주입하지 않고, App에서 메서드를 객체로 넣어주도록 수정함 (getRestaurants, onToggleFavorite)

sooyeoniya committed:
refactor(RestaurantList): DOM 초기화 로직 함수로 분리

sooyeoniya committed:
refactor: BottomSheet에 식별자 부여하여 모달 구분 및 관련 테스트 코드 수정

- 바텀 시트별로 식별자를 부여하여 클래스 속성만으로 구분하지 못해 오류나는 테스트 코드 수정함
- 추가로 사용자 정의 커맨드 메서드를 사용하여 전체 테스트 코드에서 공통으로 사용하는 로직 분리하여 적용함

sooyeoniya committed:
feat(RestaurantDetail): 상세정보 배치 수정 및 link 추가

- button 공통 스타일 분리
- link 요소 a 태그로 추가
- 필요하지 않은 스타일 정리 (category-icon)

sooyeoniya committed:
chore: 바텀시트 닫는 테스트 코드 명세 수정

sooyeoniya committed:
refactor: favorites 로컬스토리지 삭제 및 restaurants의 isFavorite로 접근하도록 수정

- 데이터 일관성 유지하기 어려움 (동기화 번거로움)
- 데이터 구조 단순화하기 위함

sooyeoniya committed:
refactor: 전체 파일 typescript 적용

sooyeoniya committed:
fix: 설명, 링크 존재하지 않을 때도 빈문자열로 무조건 값이 갱신되도록 설정

- if 조건문 때문에 이전 값이 그대로 렌더링 되는 이슈

sooyeoniya committed:
refactor: ts 변환하지 못한 파일 마저 ts 적용

sooyeoniya committed:
test: e2e 테스트 코드 작성

- 자주가는 음식점 추가 플로우
- 음식점 삭제하는 플로우
- 카테고리 필터링 및 정렬 플로우
- 상세 정보 보여주는 플로우

sooyeoniya committed:
refactor(test): 테스트 코드 리팩토링

- 재사용성을 위해 사용자 정의 메서드 사용

sooyeoniya committed:
refactor(Store): 상태 관리 로직 리팩토링

[문제점]
- 옵저버 패턴을 제대로 활용하지 못하고 있었음(리스너에 restaurantList를 보내고 있는데 이걸 사용하지 않고 그냥 직접 메서드로 받아와서 상태 변경해줌
- 필터링 정보도 해당 컴포넌트의 메서드에 접근하여 getCurrentTabType, getCurrentFilterType으로 접근했음
- 이것 때문에 컴포넌트 내부에서 service의 get메서드를 접근해 리스트를 업데이트 해주고 있었음

[해결 방식]
- 필터링 상태와 현재 선택한 음식점, 필터링된 음식점 목록을 전부 상태 관리하여 한 번에 관리해주고자 함
- 구조적으로 통일감 있고 훨씬 깔끔할 것이라 판단

- 메서드에서 값을 반환하지 않고, 메서드는 state를 업데이트 하고 해당 state를 접근해서 업데이트 해주는 방식으로 구현
- 상태 변경 시 업데이트가 필요한 restaurantList와 restaurantDetail에서 상태를 구독하여 상태가 변경될 때마다 업데이트 되도록 변경

- restaurantNavBar, restaurantFilter에서는 필터링 상태를 업데이트 해주고,
- restaurantList에서는 필터링된 정보를, restaurantDetail에서는 선택된 음식점 정보를 업데이트 해줌

sooyeoniya committed:
docs: README.md 체크박스 표시

sooyeoniya committed:
refactor(types): Category, Distance 중복 타입 정의 수정

[문제]
- 유니온 타입과 배열이 내용이 동일하여 중복 정의된 부분을 하나로 통합하고자 함

[해결책]
- 각 배열을 as const 를 사용해 각 배열 요소가 문자열 리터럴 타입으로 고정되도록 함
- typeof를 사용해 각 요소의 타입을 가져오는데 이때 number 인덱스 타입을 사
용
- 배열 내 모든 가능한 인덱스(숫자)에 대한 값 타입 추출하여 배열의 모든 오소 타입을 유니온으로 생성

sooyeoniya committed:
refactor(Store): toggleFavorite 내부에서 updateSelectedRestaurant 처리해주도록 변경

- 외부에서 메서드 두 개 접근하는 것은 불필요함 -> toggleFavorite 호출될 때 무조건 updateSelectedRestaurant 호출해야 하기 때문

wonsss commented on src/App.ts:
```diff
@@ -0,0 +1,140 @@
+    this.#renderRestaurantList();
+    this.#renderSubmitFormBottomSheet();
+    this.#renderOpenDetailBottomSheet();
+  }
```
render 로 시작하는 메서드들 모두 실행 순서에 의존적이라는 문제가 있어요.
실행 순서가 달리 되면, 아직 $body가 null이거나, $main이 null 일 수 있어 오류가 발생할 가능성이 있어요.
독립적인 렌더링을 보장하도록 개선할 필요가 있어요.

wonsss commented on src/App.ts:
```diff
@@ -0,0 +1,140 @@
+      },
+    });
+    this.$main.append(this.$restaurantNavBar.render());
+  }
```
현재 구조에서는 #renderHeader(), #renderRestaurantNavBar(), #renderRestaurantFilter() 등의 메서드가 자신이 어느 DOM 요소(this.$main, this.$body)에 추가될지 내부에서 결정되고 있어요.

render 메서드를 호출할 때 어디에 추가될지 명확히 알 수 있도록 개선해볼 수 있지 않을까요?
메서드 매개변수에 어디에 추가될지 대상을 받도록 하는 등의 방법을 고민해서요.
메서드 내부를 보지 않고도 어떤 컨테이너에 렌더링되는지 알 수 있도록 코드 가독성을 높여볼 수 있을 것 같아요.

wonsss commented on src/App.ts:
```diff
@@ -0,0 +1,140 @@
+      onOpen: () => this.$submitFormBottomSheet.open(),
+    });
+    this.$body.append($header.render());
+  }
```
현재 주어진 코드에서 좀 더 선언적으로 렌더링하는 코드를 작성하는 방법을 고민해봤어요.
현재 render 메서드들이 요소 생성(create)과 렌더(render) 두 가지 역할을 모두 하고 있어요.
이를 역할별로 분리하여 createX()는 요소를 생성하고, render()는 요소를 화면에 배치하는 방식으로 개선하면 더욱 선언적인 코드가 될 수 있어요.

"element를 container에 렌더한다"는 역할의 render(element: HTMLElement | HTMLElement[], container: HTMLElement) 같은 인터페이스의 함수를 만들어 볼 수도 있어요. 아주 간단하게 구현해봤는데 참고만 해보세요.

```ts
function render(element: HTMLElement | HTMLElement[], container: HTMLElement) {
  if (Array.isArray(element)) {
    element.forEach((child) => render(child, container));
  } else if (element instanceof HTMLElement) {
    container.appendChild(element);
  }
}
```

위와 같은 공용 render 함수를 사용하여 UI 구성 요소를 `선언적`으로 배치해볼 수 있어요. 이렇게 작성한 호출부는 예를 들어 아래와 같을 수 있어요.

```ts
  constructor(private store: RestaurantStore) {
    this.store = store;
    this.#renderComponents();
    this.#subscribeToStore();
  }

  #renderComponents() {
    this.$body = document.querySelector("body")!;
    this.$main = document.createElement("main");

    render(this.#createHeader(), this.$body);
    render(
      [
        this.#createRestaurantNavBar(),
        this.#createRestaurantFilter(),
        this.#createRestaurantList(),
        this.#createSubmitFormBottomSheet(),
        this.#createOpenDetailBottomSheet(),
      ],
      this.$main
    );
    render(this.$main, this.$body);
  }


   #createHeader(): HTMLElement { ... }
   #createRestaurantNavBar(): HTMLElement { ... }
```

의사코드 정도로 간주해주세요.
중요한 것은 함수가 한 가지 역할만 하도록 하고, 선언적으로 렌더링한다는 부분을 느끼실 수 있으면 좋을 것 같아요.

wonsss commented on src/App.ts:
```diff
@@ -0,0 +1,140 @@
+  private $restaurantList!: RestaurantList;
+  private $submitFormBottomSheet!: BottomSheetBase;
+  private $openDetailBottomSheet!: BottomSheetBase;
+  private $restaurantDetail!: RestaurantDetail;
```
`!`(Non-null 단언 연산자) 를 사용하여 요소들이 반드시 존재할 것이라고 단언하고 있는데, 그렇게 단언해도 괜찮은걸까요?
컴파일타임에서는 이처럼 단언 연산자 때문에 오류가 나지 않지만, 초기화 순서가 보장되지 않거나 실수를 하게 된다면 런타임에서 undefined 오류가 발생할 가능성이 있어요.

sooyeoniya replied:
전체 코드에서 Non-null 단언 연산자 모두 제거했습니다! 구현할 때까지만 해도 요소에 값이 무조건 들어갈 거라는 전제를 가지고 해서 그런지 사용성과 그 의미를 제대로 이해하지 않고 함부로 Non-null을 남발했던 것 같다는 생각이 들었습니다.

Non-null을 사용하면 타입 체킹 자체를 무시하게 되니까 좋은 방식이 아닌 것 같다는 생각이 들었어요! 고로 Non-null을 모두 제거하고 초기화 가능한 로직은 미리 초기화해두고, 그 외에 값은 null 값을 넣어주어 내부에서 type guard를 통해 null인 경우와 아닌 경우를 구분할 수 있도록 수정해주었습니다.
https://github.com/woowacourse/javascript-lunch/pull/257/commits/6e4f0ba90b41cdc8d7eafda0f58122de4753858c https://github.com/woowacourse/javascript-lunch/pull/257/commits/0bac712034cee356112905871f25d2583fe47dba https://github.com/woowacourse/javascript-lunch/pull/257/commits/bf07769582f94477aff2bad0103f4aa37c8ee635

wonsss commented on src/components/restaurant-filter-section/RestaurantFilter.ts:
```diff
@@ -0,0 +1,111 @@
+    };
+  }
+
+  render() {
```
모든 UI 컴포넌트가 render() 메서드들을 공통적으로 갖고 있네요.
그렇다면 이러한 인터페이스를 만들어 컴포넌트들이 implements 하여 사용하게 하면 어떨까요? 예를 들어, UIComponent 라는 인터페이스를 정의하고 모든 컴포넌트에서 이 인터페이스를 구현하도록 정의하면, 반드시 render() 메서드를 구현하게 됩니다.
이를 통해 코드 일관성이 유지되고, 타입 안정성이 증가해요.

```ts
export interface UIComponent {
  render(): HTMLElement;
}

class RestaurantFilter implements UIComponent {

```

sooyeoniya replied:
이 부분에 대해서 전적으로 동의합니다! 작성해주신 코드를 참고하여 UIComponent 인터페이스 정의해서 컴포넌트에 대한 타입 안정성을 높여주었습니다. https://github.com/woowacourse/javascript-lunch/pull/257/commits/637500a898386c2d7617713d39a333023fb6e24d

wonsss commented on src/stores/RestaurantStore.ts:
```diff
@@ -0,0 +1,142 @@
+    this.#loadFromLocalStorage();
+  }
+
+  get state(): StoreState {
```
프라이빗 필드와 getter 사용을 잘하셔서 캡슐화가 잘 되어 있네요!

wonsss commented on src/stores/RestaurantStore.ts:
```diff
@@ -0,0 +1,142 @@
+      sortFilterType: "name",
+    },
+  };
+  #selectedRestaurant: Restaurant = {} as Restaurant;
```
타입단언은 가능한 사용하지 않는 게 좋아요. 타입단언을 쓰게 되면 타입안정성이 깨질 수 있어요.
현재 selectedRestaurant의 초기값이 {}(빈 객체)로 설정되어 있는데, 논리적으로 선택된 음식점이 없을 경우 null이 더 적절할 것 같아요.
그러나 개발의 편의를 위해 타입 단언(as Restaurant)을 사용하면, 이후 코드에서 selectedRestaurant가 undefined일 가능성을 무시하게 만들어 런타임 버그의 원인이 될 수 있어요.

```
#selectedRestaurant: Restaurant | null = null;
```

sooyeoniya replied:
반영했습니다, 감사합니다 :) 자세한 사항은 커밋 내용에 작성해두었습니다! https://github.com/woowacourse/javascript-lunch/pull/257/commits/8bca4c6719f9ab380e3e6203ed9f7080c2826fcc

wonsss commented on src/stores/RestaurantStore.ts:
```diff
@@ -0,0 +1,142 @@
+      },
+      [LABEL_KEYS.distance]: (restaurantsInfo: Restaurant[]) => {
+        return restaurantsInfo.sort(
+          (a, b) => parseInt(a.distance) - parseInt(b.distance)
```
현재 sort() 메서드는 배열을 정렬하면서 원본 배열을 직접 변경하고 있어요.
이로 인해 불변성이 깨지고, 예상치 못한 사이드이펙트가 발생할 가능성이 있어요.

sort()를 사용하기 전에 slice() 나 스프레드 연산자 등의 방법으로 배열을 복사하여 원본 데이터를 보호할 수 있어요.

> 참고
<img width="1210" alt="image" src="https://github.com/user-attachments/assets/ace22432-7335-4226-8765-fc8f170b1f00" />
https://poiemaweb.com/js-array-higher-order-function

sooyeoniya replied:
`tabTypeFn`와 `sortFilterTypeFn` 이 두 객체를 사용할 때 `getFilteredRestaurants` 메서드 내부에서 이미 처음에 원본 배열을 한 번 복사해서 새로운 배열을 사용해 검증하고 있기 때문에 불변성 문제는 발생하진 않을 것 같습니다!

```jsx
const restaurants = [...this.#restaurants];

...

if (tabType === NAV_BAR_KEYS.favorite) {
  return tabTypeFn[tabType](restaurants); // 복사된 restaurants 사용
}

return sortFilterTypeFn[sortFilterType](tabTypeFn[tabType](restaurants));
```

만약 이 두 객체가 다른 곳에서도 사용된다면 객체 내부에서 새로운 배열을 만들도록 스프레드 연산자를 사용해서 복사해주면 좋을 것 같습니다. 

하지만, 현재 상황에서는 한 번 더 스프레드로 감싸주게 되면 중복 배열 복사가 일어나서 불필요한 메모리를 낭비하는 것 같다는 생각이 들어서 수정하지 않기로 했습니다!

wonsss commented on types/interfaces.ts:
interface와 type은 어떤 차이가 있나요?

> 참고 https://yceffort.kr/2021/03/typescript-interface-vs-type

sooyeoniya replied:
이전에 학습했던 내용이 기억이 잘 나지 않아서 다시 공부해봤습니다!

**1. 확장 문법차이**
- Interface: extends 키워드 사용
- Type: & 연산자 사용

**2. 선언 병합**
- Interface: 가능 (같은 이름으로 여러 번 선언 시 자동 병합)
- Type: 불가능 (중복 선언하면 에러 발생함)

**3. Union, Intersection 타입**
- Interface: 직접 유니온 및 인터섹션 선언 불가
- Type: 유니온 및 인터섹션 타입 선언 가능 (인터페이스에 있는 걸 대신 유니온/인터섹션 가능)

**4. 계산된 프로퍼티: 유연함 차이**
- Interface: 매핑된 타입 직접 사용 불가, 유니온 타입과 조합 불가, 재귀적 타입 사용 불가
- Type: 매핑된 타입 사용 가능, 유니온 타입과 조합 가능, 재귀적 타입 사용 가능 (더 유연하게 사용 가능)

### 결론
간단한 객체 구조를 정의할 때는 Interface쓰고, 유니온 타입이나 타입 변환이 새롭게 필요한 경우에는 Type 쓰는 것이 좋을 듯 합니다!

sooyeoniya replied:
참고 자료도 천천히 다 읽어봤습니다! 좋은 정보 공유 감사합니다 😃

wonsss commented on src/stores/RestaurantStore.ts:
```diff
@@ -0,0 +1,142 @@
+    this.#notifyListeners();
+  }
+
+  getFilteredRestaurants({
```
getFilteredRestaurants 는 Store가 아니라 UI에서 하는 것이 더 자연스러울 것 같아요.
Store는 음식점 데이터만 관리하고, UI에서 필요할 때 필터링을 수행하는 게 유지보수에 좋을 것 같아요.

sooyeoniya replied:
확실히 `Store`가 지금 너무 많은 책임을 담당하고 있기도 하고, 정렬이나 필터링 같은 경우에는 단순히 `restaurants`와 `currentFilter` 이 두 개의 상태를 조합해서 만드는 것이기 때문에 UI로 책임분리할 수 있는 방법이 없는지 좀 더 고민해봤습니다.

현재는 `getFilteredRestaurants` 메서드 자체가 `restaurantList` 컴포넌트에 있는 `updateRestaurantList` 메서드 즉, 음식점 목록을 필터링된 목록으로 UI를 업데이트 해주는 로직에서만 사용이 되고 있습니다. 따라서, `Store`에 있는 상태 중 `filteredRestaurants(필터링된 음식점 목록들) 값 자체를 제거하는 것`이 더 좋은 방향이라고 생각했습니다. 

이렇게 하게 되면 `Store`에서는 정말 필요한 데이터만 가지고 있게 되고, 음식점 목록 필터링이 필요한 UI 부분에서는 `getFilteredRestaurants` 메서드를 사용해 바로바로 업데이트 해줄 수 있기 때문에 좀 더 유연하게 데이터를 가공할 수 있다는 생각이 들었습니다.

따라서 `getFilteredRestaurants를 아예 utils로 분리`해주기로 했습니다! 사실 이렇게 되면 이 메서드가 외부 상태에 의존하고 있고, 어딘가에 종속되어있는 메서드가 아니기 때문에 책임소재가 약간 불분명해지는 건 사실입니다. 하지만 오히려 재사용 하는데 좀 더 수월해지고 필터링 로직을 한 곳에서 관리할 수 있어서 훨씬 더 장점이 많다는 생각이 들었습니다. 따라서 `getFilteredRestaurants` 는 유틸보다는 헬퍼함수에 가깝다고 보면 될 것 같습니다.

추가로 맨처음 App에서 음식점 목록을 렌더링 해줄 때 `getFilteredRestaurants` 를 사용하는데요. 이때는 필터링 옵션 자체를 넘겨주지 않고 기본 필터링 옵션 값으로 설정되도록 하였습니다. 이때 기본 값은 `“모든 음식점 보기”, “전체”, “이름순”` 입니다.

따라서, App에서 처음 음식점 목록을 렌더링 할 때는 기본값으로 필터링이 설정되도록 했고, 추후 현재 필터링 옵션 값으로 음식점 목록을 처리할 때는 필터링 값을 넘겨 그 값으로 필터링된 음식점 목록을 반환하도록 해주었습니다.
https://github.com/woowacourse/javascript-lunch/pull/257/commits/b5f3aa387440e8bfd38078cb93b5b62ca78b7db8

wonsss replied:
잘 수정해주셨네요!

wonsss commented on src/stores/RestaurantStore.ts:
```diff
@@ -0,0 +1,142 @@
+    return () => this.#listeners.delete(key);
+  }
+
+  #loadFromLocalStorage() {
+    const savedRestaurants = localStorage.getItem("restaurants");
+    if (savedRestaurants) {
+      this.#restaurants = JSON.parse(savedRestaurants);
+      this.#notifyListeners();
+    }
+  }
+
+  #saveToLocalStorage() {
+    localStorage.setItem("restaurants", JSON.stringify(this.#restaurants));
+  }
```
로컬 스토리지 로직은 별도 모듈로 분리하여 재사용 가능하게 만들 수 있을 것 같아요.

sooyeoniya replied:
https://github.com/woowacourse/javascript-lunch/pull/257/commits/d913d9a8f116305b59c99d90991b4824c70d6b35
이렇게 적용해봤습니다! get, set, remove, clear 메서드를 정의하는 Storage라는 인터페이스를 생성하고 이걸 LocalStorage라는 로컬스토리지를 위한 클래스로 확장시켜 로컬스토리지 재사용을 위한 클래스를 만들어주었습니다. 약간은 오버엔지니어링 같은 느낌이 있지만, 확장성을 위해 이런 방식으로 리팩토링 해보고 싶었습니다. 이렇게 하면 추후 Storage 라는 인터페이스로 세션 스토리지 같은 클래스를 추가로 만들 수 있을 것 같다는 생각이 듭니다.

그리고 LocalStorage 를 조합(composition)해서 RestaurantStorage를 만들어주었는데요. 그 이유는 `RestaurantStorage`가 `LocalStorage`에 대해 `is-a` 관계가 아닌 `has-a` 관계이기 때문에 좀 더 느슨한 결합을 위해 상속보다는 조합이 더 적합하다는 생각을 했습니다. 그래서 내부에서 LocalStorage 인스턴스를 생성하여 메서드를 구현해주었습니다.

wonsss commented on src/stores/RestaurantStore.ts:
```diff
@@ -0,0 +1,142 @@
+    return sortFilterTypeFn[sortFilterType](tabTypeFn[tabType](restaurants));
+  }
+
+  subscribe(key: string, callback: (state: StoreState) => void) {
+    this.#listeners.set(key, callback);
+    callback(this.state);
+
+    return () => this.#listeners.delete(key);
+  }
```
현재 RestaurantStore 가 옵저버 패턴 자체를 직접 관리하고 있는데, 만약 옵저버 패턴을 다른 곳에서 사용하려고 하면 재사용이 어려워요. 또한 옵저버 패턴을 포함하면서 코드 복잡도와 가독성이 떨어질 수 있어요.

옵저버 패턴도 다른 Store에서 활용 가능하도록 일반화하여 분리해볼 수 있을 것 같아요.

sooyeoniya replied:
https://github.com/woowacourse/javascript-lunch/pull/257/commits/1d39fed57fc799a5d5ae55ba5dabc1f8b6892726
옵저버 패턴에 필요한 `subscribe`와 `notify` 부분을 재사용하기 위해 추상 클래스로 정의하여 상속해 사용해주었습니다! 이때 상태에 대한 타입을 제네릭으로 받아서 처리되도록 해주었습니다. 

그리고 기존에 `private`였던 `listeners` 나 `notify` 같은 경우에는 상속 과정에서 기존 접근제한자가 바뀌지 않도록 `protect`로 변경하였습니다.

wonsss commented on cypress/e2e/filterAndSortRestaurants.cy.js:
```diff
@@ -0,0 +1,138 @@
+  });
+
+  describe("음식점 목록에서 사용자가 원하는 카테고리로 필터링 시 해당 카테고리에 대한 음식점 목록만 보인다.", () => {
+    const testCases = [
+      {
+        category: "전체",
+        visibleRestaurants: ["얌샘 김밥", "메콩타이"],
+        hiddenRestaurants: [],
+      },
+      {
+        category: "한식",
+        visibleRestaurants: ["얌샘 김밥"],
+        hiddenRestaurants: ["메콩타이"],
+      },
+      {
+        category: "아시안",
+        visibleRestaurants: ["메콩타이"],
+        hiddenRestaurants: ["얌샘 김밥"],
+      },
+    ];
+
+    testCases.forEach(({ category, visibleRestaurants, hiddenRestaurants }) => {
```
testCases  배열을 사용하여 반복적인 테스트를 줄이고 가독성을 높인 점이 좋네요 👍

wonsss commented on cypress/support/commands.ts:
Cypress 의 커스텀 커맨드 기능을 적극적으로 잘 사용해주셨네요 👍

wonsss reviewed (changes requested):
안녕하세요 니야! 2단계 미션 수고 많으셨어요.
타입스크립트 적용 잘 해주셨고, 설계적으로 수정한 부분도 더 좋아졌네요.
많은 고민을 하신 것 같아 멋집니다. 👍 
리뷰 남겨드린 내용 참고해주세요~!

sooyeoniya committed:
refactor(App): Non-null 단언 연산자 제거 및 undefined 설정

[문제점]
- Non-null 단언 연산자를 사용할 경우 타입 체킹 과정에서 무시되기 때문에 런타임 과정에서 undefined 될 가능성 있음

[해결책]
- Non-null 단언 연산자 제거 후 undefined 타입을 명시해주고, if 문으로 type guard 해줌
- body 의 경우에는 document.body 로 접근하여 추론된 타입을 바꿔줌
- DOMContentLoaded 이벤트를 걸어서 HTML를 읽어 DOM 트리가 전부 완성된 후에 코드 실행되도록 함 => body가 undefined될 가능성X

[body 요소 접근 방법을 다르게 해준 이유]
- document.body와 document.querySelector('body')는 서로 타입이 다르게 지정됨
- document.body의 반환타입은 HTMLElemet로 HTML 문서에 항상 body가 존재하므로 null일 가능성이 x
- document.querySelector('body')의 반환타입은 HTMLBodyElement | null 로 CSS 선택자로 요소를 찾기 때문에 없으면 null 가능성 O

sooyeoniya committed:
refactor: 요소 append 로직과 create 로직 분리

- appendElemnt 메서드 분리하여 재사용하도록 수정
- render 보다는 append 가 해당 메서드의 역할을 잘 드러낸다고 판단하여 appendElement로 결정함

sooyeoniya committed:
refactor: 컴포넌트 인스턴스 생성 로직 분리

- 관심사 분리 및 재사용성 높이기
- 선언형 패러다임 적용

sooyeoniya committed:
refactor: export 파일 모두 index에서 export 되도록 수정

sooyeoniya committed:
refactor: store 객체 생성을 App 내부로 이동

- 해당 App에서만 사용되기 때문에 응집도 높이기 위함

sooyeoniya committed:
refactor: store 구독 로직 분리

- store 내부에서 subscribe 관리함
- 여러 store 에 대한 subscribe 가 생성될 수 있도록 확장성 고려하여 함수에 Restaurant 붙여서 정의

sooyeoniya committed:
refactor: 이벤트 핸들러 메서드 분리

- 외부로 분리하여 재사용할 수 있도록 함

sooyeoniya committed:
refactor: open-detail -> detail, submit-form -> submit로 변경

- open은 동사이기 때문에 상세정보 바텀시트 컴포넌트 명칭이 올바르지 않다고 판단함
- 따라서, 'open' prefix 제거

sooyeoniya committed:
chore: appendElement -> render로 메서드명 변경

- 전체 컴포넌트의 UI 렌더링 관련하여 render 메서드를 사용하고 있기 때문에 일관된 패턴 유지를 위해 변경

sooyeoniya committed:
refactor(RestaurantDetail): non-null 단언 연산자 제거 및 내부 리팩토링

- non-null 단언 연산자 제거 및 초깃값 설정
- 어차피 RestaurantDetail이 생성되는 시점에 요소들이 같이 생성되므로 초기화 부분 옮김
- 음식점 정보에 대해 전부 this로 접근하지 않고 구조분해할당하여 값 업데이트
- id 의 경우에는 다른 내부 메서드에서도 사용되기 때문에 this로 저장

sooyeoniya committed:
refactor: 전체 파일 non-null 단언 연산자 제거

sooyeoniya committed:
refactor(RestaurantStore): selectedRestaurant의 타입 및 초깃값 변경

- 빈 객체에 타입 단언을 하면 런타임에서 undefined 오류 발생 가능
- 따라서, '선택되지 않음' 상태를 명확히 표시하기 위해 null 값으로 초기화
- state를 get으로 외부에서 접근 시 selectedRestaurant가 존재하지 않는 경우 객체에 대한 초깃값 주입

sooyeoniya committed:
refactor(RestaurantDetail): this.id에 대한 타입 가드 추가

- null이 아닌 경우에만 this.id 를 사용할 수 있도록 수정

sooyeoniya committed:
refactor: UIComponent 인터페이스 정의 및 적용

- 각 컴포넌트의 공통 render 메서드의 반환 타입 정의
- 타입 일관성 및 안정성 향상, render메서드가 필수로 구현되도록 통일

sooyeoniya committed:
refactor: 필터링된 음식점 목록에 대한 로직 Store에서 분리

- Store 에는 음식점 목록과, 필터링 옵션 값들, 선택된 음식점에 대한 정보만 관리하도록 수정
- 필터링된 음식점 목록은 Store에서 관리하는 것보다 필터링된 음식점 목록이 필요한 부분에서 사용하여 UI 업데이트 해주도록 함
- getFilteredRestaurants 메서드에서 필터링 옵션에 대한 기본값이 설정되도록 함
- 처음 필터링된 음식점 목록을 UI에 렌더링할 때는 가장 기본값인 '모든 음식점 보기', '전체', '이름순'으로 설정

sooyeoniya committed:
refactor: 로컬스토리지 로직 모듈화

- 스토리지 인터페이스 생성하여 해당 인터페이스를 사용하는 로컬스토리지 클래스 생성
- 해당 로컬스토리지 클래스를 사용하는 RestaurantStorage 클래스 생성
- 내부에서 해당 키와 타입 관리하도록 함
- 확장성 고려를 위해 모듈화함 -> e.g., 세션스토리지 같은 추상클래스도 추가 가능

sooyeoniya committed:
refactor: 옵저버 패턴 추상 클래스로 정의

- 옵저버 패턴에 주요 메서드인 subscribe와 notify 로직 추상화하여 재사용성 향상
- RestaurantStore에서 상속받아 제너럴 타입에 StoreState 를 넘겨 사용하도록 함
- 구독하는 로직 리팩토링: 두 개의 컴포넌트에 대한 구독을 분리함. 인자로 넘길 때 컴포넌트 부분을 객체로 묶어서 넘김(store와 구분짓기 위함)
- 본 프로젝트에서는 unsuscribe를 사용하진 않지만 반환값이 있으므로 배열에 저장해두도록 수정

sooyeoniya committed:
refactor: stores 내부 폴더 분리

sooyeoniya committed:
docs: README.md 폴더 구조 수정

sooyeoniya commented:
## ✅ App 내부 리팩토링

### ✔️ 목표

- 선언형 프로그래밍하기
- non-null 단언 연산자 제거하기
- null 타입 가드 추가하기

### ✔️ render 메서드 생성 (특정 요소에 append 해주는 메서드)

우선 render로 시작하는 메서드마다 내부에 모두 append 해주는 로직이 있었는데요. 이 부분 때문에 마르코가 실행 순서에 의존적이라고 분리하라는 피드백이 있었습니다. 그래서 이 부분을 개선하기 위해 append 해주는 로직을 모두 분리하여 한 번에 요소에 append 해줄 수 있는 render() 라는 유틸 함수를 하나 생성해주었습니다. 마르코가 주신 의사코드를 참고해서 구현해봤어요! 감사합니다 ㅎㅎ

덕분에 initializeCompoenents 라는 메서드에서 한 번에 실행 순서를 관리할 수 있어서 이전 코드보다 훨씬 유지보수하기 쉬워진 것 같습니다.

### ✔️ $body와 $main이 Null일 경우 고려

그리고 $body가 null이거나 $main이 null인 경우를 모두 확인해서, 둘 다 null이 아닌 경우에만 render() 메서드를 사용할 수 있도록 해주었습니다. $body 같은 경우에는 index.html에 존재하기 때문에 아예 JS 파일이 실행 로직에 DOMContentLoaded 이벤트를 걸어주어 HTML이 모두 로딩된 후에 JS 파일이 실행되도록 해주었습니다. 이렇게 해서 $body 가 존재하지 않을 가능성을 없애주었습니다.

### ✔️ builder와 handler 분리

선언형 패러다임과 더불어 관심사를 분리하기 위해서 어떻게 더 리팩토링할 수 있을지 고민해봤습니다. 그 과정에서 고안해낸 방법은 App에 있는 로직에서 builder와 handler를 따로 분리하는 것이었습니다.

`builder`는 각 컴포넌트 인스턴스를 생성하여 반환해주는 모듈입니다. 사실 `BottomSheetBase` 컴포넌트 같은 경우에는 제출 폼에서 한 번, 상세 정보에서 한 번, 이렇게 총 2번 인스턴스가 생성되었는데요. builder로 분리함으로써 재사용할 수 있게 되었습니다. 그리고 컴포넌트 생성 로직을 해당 builder 모듈에서 관리하면 되기 때문에 가독성 측면에서나 유지보수 측면에서도 훨씬 좋아진 것 같다는 생각이 듭니다.

`handler`의 경우에는 이벤트 콜백 함수를 모아둔 모듈인데요. 이런 handler들도 `handleCloseBottomSheet`나 `handleToggleFavorite` 같이 동일한 메서드가 구현되어 있는 경우가 있었습니다. 이에 대해 재사용성을 높이고자 이벤트 핸들러를 한 곳에 모아두기로 결정했습니다.

원래는 하나의 컴포넌트에 대하여 builder(컴포넌트 생성)와 handler(이벤트 처리)를 객체나 클래스로 한 번에 관리하려고 했는데, 여러 컴포넌트에서 재사용되는 로직이 있어서 지금과 같이 두 개로 관심사를 분리해봤습니다.

이렇게 수정한 덕분에 App에서 봤을 때 좀 더 선언적인 코드가 된 것 같아서 훨씬 가독성이 좋아졌다는 생각이 듭니다.

사실 이렇게 했을 때 문제점도 많이 보여서 약간의 트레이드오프 현상이 발생하는 것 같다는 생각이 들었습니다. builder와 handler로 분리하면 이것만의 장점이 있기도 하지만, 실질적으로 분리만 되었을 뿐 App과의 결합도가 높아서 이 부분에 있어서 약간 책임이 분산되는 경향이 있는 것 같습니다.

차라리 이런 방식으로 분리하지 않고, 앞서 생각했던 다른 방식인, 각 컴포넌트 단위로 묶어서 컴포넌트 생성과 그 컴포넌트에 대한 이벤트 처리를 한 번에 통합하는 것이 더 나은 방식일지 고민이 됩니다. 이렇게 되면 훨씬 응집도가 높아질 것 같다는 생각도 드네요.

wonsss reviewed (approved):
안녕하세요 니야! 😄 
전체적으로 선언형 프로그래밍, 관심사 분리, 안정성 향상(null 타입 가드) 등을 명확한 목적 의식과 함께 의도적으로 개선한 점이 인상 깊어요. 피드백에 따라 단순히 '작동하는 코드'를 넘어서 읽기 좋고, 관리하기 쉬운 구조로 발전시키려는 노력하신 것 같아 좋았어요. 또한, 각각의 개선사항마다 “왜 그렇게 했는지” 명확한 이유와 고민의 흔적도 남겨주셔서 좋았습니다. 
수고 많으셨어요. 앞으로도 화이팅입니다! 💪
